FastSimulation/Calorimetry/src/KKCorrectionFactors.cc

```diff
@@ -0,0 +1,75 @@
+#include "FastSimulation/Calorimetry/interface/KKCorrectionFactors.h"
+
+KKCorrectionFactors::KKCorrectionFactors( const edm::ParameterSet& pset )
+{
+
+  interpolate_ = pset.exists("interpolate") && pset.getUntrackedParameter<bool>("interpolate");
+
+  // Get the filename and histogram name
+  std::string ecalScalesFileName_ = pset.getUntrackedParameter<std::string>("fileName");
+  std::string ecalScalesHistogramName_ = pset.getUntrackedParameter<std::string>("histogramName");
+
+  // Read histo
+  edm::FileInPath myDataFile( ecalScalesFileName_ ); // TODO: check if this is necessary
+  TFile * myFile = new TFile( myDataFile.fullPath().c_str(), "READ" );
+
+  gROOT->cd(); // create histogram in memory
+  auto obj = myFile->Get( ecalScalesHistogramName_.c_str() );
+
+  // Check if histogram exists in file
+  if(!obj) {
+    throw cms::Exception( "FileReadError" )
+      << "Histogram \"" << ecalScalesHistogramName_
+      << "\" not found in file \"" << ecalScalesFileName_
+      << "\", used for correcting the response of the ECAL in FastSim.\n";
+  }
+  h3_ = new TH3F( *((TH3F*)obj) );
+
+  // fill empty bins with 1.
+  for( int x=0; x<h3_->GetNbinsX()+2; x++ ) {
+    for( int y=0; y<h3_->GetNbinsY()+2; y++ ) {
+      for( int z=0; z<h3_->GetNbinsZ()+2; z++ ) {
+        if( h3_->GetBinContent(x,y,z) < 1e-5 ) {
+          h3_->SetBinContent(x,y,z,1.);
+        }
+      }
+    }
+  }
+
+  myFile->Close();
+  delete myFile;
```

Comment:
Ah, I don't know why I had that carved the wrong way in my brain.

Googled a bit.
Deleting is necessary indeed,
but deleting includes closing,
so no need to close here before deleting
https://root.cern.ch/root/roottalk/roottalk03/5385.html

On Wed, Aug 19, 2015 at 2:33 PM, Chris Jones <notifications@github.com>
wrote:

> In FastSimulation/Calorimetry/src/KKCorrectionFactors.cc
> <https://github.com/cms-sw/cmssw/pull/10596#discussion_r37408546>:
>
> > +  }
> > +  h3_ = new TH3F( *((TH3F*)obj) );
> > +
> > +  // fill empty bins with 1.
> > +  for( int x=0; x<h3_->GetNbinsX()+2; x++ ) {
> > +    for( int y=0; y<h3_->GetNbinsY()+2; y++ ) {
> > +      for( int z=0; z<h3_->GetNbinsZ()+2; z++ ) {
> > +        if( h3_->GetBinContent(x,y,z) < 1e-5 ) {
> > +          h3_->SetBinContent(x,y,z,1.);
> > +        }
> > +      }
> > +    }
> > +  }
> > +
> > +  myFile->Close();
> > +  delete myFile;
>
> We still need a delete, even with TFile::Open.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10596/files#r37408546>.
>